util/design/lib/OtpMemImg.py

```diff
@@ -0,0 +1,384 @@
+        # validation and image generation depends on them
+        self.lc_state = LcStEnc(lc_state_config)
+
+        # Validate memory image configuration
+        log.info('')
+        log.info('Parse OTP image specification.')
+
+        # Encryption sanity test with known test vector
+        enc_test = _present_64bit_encrypt(
+            int("0123456789abcdef", 16),
+            int("0123456789abcdef0123456789abcdef", 16))
+
+        if enc_test == int('0e9d28685e671dd6', 16):
+            log.info('Encryption module test OK')
+        else:
+            raise RuntimeError('Encryption module test failed')
+        log.info('')
+
+        if 'seed' not in img_config:
+            raise RuntimeError('Missing seed in configuration.')
+
+        img_config['seed'] = check_int(img_config['seed'])
+        log.info('Seed: {0:x}'.format(img_config['seed']))
+        log.info('')
+
+        # Re-initialize with seed to make results reproducible.
+        random.seed(OTP_IMG_SEED_DIVERSIFIER + int(img_config['seed']))
+
+        if 'partitions' not in img_config:
+            raise RuntimeError('Missing partitions key in configuration.')
+
+        for part in img_config['partitions']:
+            self.merge_part_data(part)
+            log.info('Adding values to {} partition.'.format(part['name']))
+            for item in part['items']:
+                self.merge_item_data(part, item)
+
+        log.info('')
+        log.info('Parsing OTP image successfully completed.')
+
+    def merge_part_data(self, part):
+        '''This validates and merges the partition data into the memory map dict'''
+
+        part.setdefault('lock', 'false')
+        part['lock'] = check_bool(part['lock'])
+        part.setdefault('items', [])
+
+        if part.setdefault('name', 'unknown_name') == 'LIFE_CYCLE':
+            part.setdefault('state', 'RAW')
+            part.setdefault('count', 'RAW')
+            part['count'] = str(part['count'])
+            if len(part['items']) > 0:
+                raise RuntimeError(
+                    'Life cycle items cannot directly be overridden')
+            if part['lock']:
+                raise RuntimeError('Life cycle partition cannot be locked')
+
+            # Augment life cycle partition with correct life cycle encoding
+            state = self.lc_state.encode('lc_state', part['state'])
+            count = self.lc_state.encode('lc_cnt', part['count'])
+            part['items'] = [{
+                'name': 'LC_STATE',
+                'value': '0x{:X}'.format(state)
+            }, {
+                'name': 'LC_TRANSITION_CNT',
+                'value': '0x{:X}'.format(count)
+            }]
+        else:
+            if len(part['items']) == 0:
+                log.warning("Partition does not contain any items.")
+
+        # Check if partition name exists in memory map and get index
+        part_idx = self.get_part_idx(part['name'])
```

Comment:
This would probably be cleaner if `self.get_part_idx` (renamed?) also returned `self.config['partitions'][part_idx]`. Indeed, I don't think you actually need `part_idx` or at any of the other call sites - searching for `part_idx` just finds hits for `self.config['partitions'][part_idx]`. So maybe it should be something like `self.get_partition` which returns `None` if there's no match.

Reply:
no but you are right that this can be confusing.
I've modified these functions to just return the partitions/items directly instead of the indices.